foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/sessions/UnitOfWorkImpl.java

```diff
@@ -5000,6 +5019,62 @@ public void setMergeManager(MergeManager mergeManager) {
      */
     protected void setNewObjectsCloneToOriginal(Map newObjects) {
         this.newObjectsCloneToOriginal = newObjects;
+        setupPrimaryKeyToNewObjects();
+    }
+
+    /**
+     * INTERNAL:
+     * Create the map of primary key to new objects used to speed up in-memory querying.
+     */
+    protected void setupPrimaryKeyToNewObjects() {
+        primaryKeyToNewObjects = null;
+        if (hasNewObjects()) {
+            primaryKeyToNewObjects = new HashMap<>();
+            getNewObjectsCloneToOriginal().forEach((object, o2) -> {
+                addNewObjectToPrimaryKeyToNewObjects(object);
+            });
+        }
+    }
+
+    /**
+     * INTERNAL:
+     * Extracts the primary key from a new object and puts it in primaryKeyToNewObjects.
+     */
+    protected void addNewObjectToPrimaryKeyToNewObjects(Object newObject){
+        ClassDescriptor descriptor = getDescriptor(newObject.getClass());
+        addNewObjectToPrimaryKeyToNewObjects(newObject,descriptor);
+    }
+
+    /**
+     * INTERNAL:
+     * Extracts the primary key from a new object and puts it in primaryKeyToNewObjects.
+     * Allows passing of the ClassDescriptor.
+     */
+    protected void addNewObjectToPrimaryKeyToNewObjects(Object newObject, ClassDescriptor descriptor){
+        ObjectBuilder objectBuilder = descriptor.getObjectBuilder();
+        Object pk = objectBuilder.extractPrimaryKeyFromObject(newObject, this, true);
+        if(pk!=null) {
+            getPrimaryKeyToNewObjects().putIfAbsent(pk, new ArrayList<>());
+            getPrimaryKeyToNewObjects().get(pk).add(newObject);
+        }
+    }
+    /**
+     * INTERNAL:
+     * Extracts the primary key and removes an object from primaryKeyToNewObjects.
+     */
+    protected void removeObjectFromPrimaryKeyToNewObjects(Object object){
+        ClassDescriptor descriptor = getDescriptor(object.getClass());
+        ObjectBuilder objectBuilder = descriptor.getObjectBuilder();
+        Object pk = objectBuilder.extractPrimaryKeyFromObject(object, this, true);
+        removeObjectFromPrimaryKeyToNewObjects(object,pk);
+    }
+
+    /**
+     * INTERNAL:
+     * Removes an object from primaryKeyToNewObjects.
+     */
+    protected void removeObjectFromPrimaryKeyToNewObjects(Object object, Object primaryKey){
+        getPrimaryKeyToNewObjects().getOrDefault(primaryKey, new ArrayList<>(0)).remove(object);
```

Comment:
Is it needed to keep in `Map<Object, List<Object>> primaryKeyToNewObjects;` record when `List<Object>` will be empty? Maybe extend removal there to whole Map records.

Reply:
I've changed it, so that the `List<Object>` is removed when it's empty.